### PR TITLE
Ripout gogo

### PR DESCRIPTION
**Pull request by EItanya**
# Description

This PR fully removes `gogoproto` as the main engine of code generation, and proto tooling, and replaces it with standard `golang` `proto`. There are a few main parts of this work:

1. Removing all `gogoproto/gogo.proto` references, and extensions from protos. The important ones, such as `deep_equal`, and `sensitive_logging` have been replaced with functionality in `protoc-gen-ext`.
2. Switching all `proto` imports to use the full `GOPATH` style importing to fix code-gen issues with the new `protoc-gen-go`
3. Switching all copied envoy protos to use the package prefix `solo.io.` to solve an issue where `protoc-gen-go` cannot register multiple types of the same `package.MessageName`.
4. Making sure to not use `jsonpb` or any other utilities from `gogo/protobuf`, as they will lead to incorrect, or crashing behavior. Despite the lack of compiler errors.
5. Using proto `gomega` and `gomock` matchers where appropriate and necessary, as the new `golang` `proto` has a habit of modifying the internal state of it's messages.

# Context

`solo-kit`, and by extension `Gloo` use `gogoproto` as their code generation mechanism. However, that project is no longer supported, so it's best to move away from it. Also, `envoyproxy/go-control-plane` migrated away from `gogoproto` to `golang`, and there are breaking changes, which means at the current moment we cannot pull in the latest envoy API, which we need.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/external/envoy/annotations/deprecation.proto [hunk 1]
```diff
@@ -12,14 +12,16 @@ import "google/protobuf/descriptor.proto";
 // Magic number in this file derived from top 28bit of SHA256 digest of
 // "envoy.annotation.disallowed_by_default"
 extend google.protobuf.FieldOptions {
-  bool disallowed_by_default = 189503207;
+  bool disallowed_by_default = 189503208;
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gloo/api/external/envoy/annotations/deprecation.proto:15]: If it matches the upstream one proto will get mad. Options have to be globally unique :(

File: projects/gateway/pkg/translator/merge_route_plugins_test.go [hunk 2]
```diff
@@ -3,11 +3,11 @@ package translator
 import (
 	"time"
 
+	"github.com/golang/protobuf/ptypes/wrappers"
+	prototime "github.com/libopenstorage/openstorage/pkg/proto/time"
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gateway/pkg/translator/merge_route_plugins_test.go]: Discussed and swap this out for solo-kit instead

File: pkg/utils/protoutils/utils.go [hunk 3]
```diff
@@ -43,15 +39,18 @@ func MarshalStructEmitZeroValues(m proto.Message) (*structpb.Struct, error) {
 	return &pb, err
 }
 
-func UnmarshalStruct(structuredData *structpb.Struct, into proto.Message) error {
+func UnmarshalStruct(structuredData *structpb.Struct, into interface{}) error {
```
[the remark this answers is not in the document]
Comment (EItanya) [reply pkg/utils/protoutils/utils.go:42]: Gonna get rid of this as it is a clone of the solo-kit version, and only used as such

File: projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto [hunk 4]
```diff
@@ -2,17 +2,16 @@
 
 syntax = "proto3";
 
-package envoy.extensions.filters.http.wasm.v3;
+package solo.io.envoy.extensions.filters.http.wasm.v3;
 
 import "envoy/extensions/wasm/v3/wasm.proto";
 
 import "validate/validate.proto";
 // manually removed udpa annotations and added gogo equal/hash, go_package
```
Comment (sam-heilbron) [inline projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto:10]: Leftover reference to gogo?

File: projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto [hunk 5]
```diff
@@ -2,17 +2,16 @@
 
 syntax = "proto3";
 
-package envoy.extensions.filters.http.wasm.v3;
+package solo.io.envoy.extensions.filters.http.wasm.v3;
 
 import "envoy/extensions/wasm/v3/wasm.proto";
```
Comment (sam-heilbron) [inline projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto:7]: Why is this import left unchanged but the other envoy external references are all prepended with `github.com/solo-io/gloo/projects/gloo/api/external/`

**Review by sam-heilbron — COMMENTED** [review]
This largely looks awesome! A few smalls nits/questions.

I think there's also a reference to gogo in the setting up a dev environment doc: https://github.com/solo-io/gloo/blob/ripout-gogo/docs/content/guides/dev/setting-up-dev-environment.md

Once this change goes into master, will this lock solo-projects into using this version of gloo until we ripout gogo as well?

File: projects/gateway/pkg/reporting/check_source_reports.go [hunk 6]
```diff
@@ -12,7 +12,7 @@ func AllSourcesAccepted(reports reporter.ResourceReports, configObj translator.O
 	allSourcesAccepted := true
 
 	if err := translator.ForEachSource(configObj, func(src translator.SourceRef) error {
-		_, report := reports.Find(src.ResourceKind, core.ResourceRef{src.Name, src.Namespace})
+		_, report := reports.Find(src.ResourceKind, &core.ResourceRef{Name: src.Name, Namespace: src.Namespace})
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gateway/pkg/reporting/check_source_reports.go:15]: Gogo allowed for non pointer struct values, as well as the ability to remove special `proto` fields from the proto structs. This combination allowed us to use the `core.ResourceRef` struct as the key for maps as well as many other uses. However, golang proto does not have either of these features, so `core.ResourceRef` will always be a pointer.

File: pkg/utils/protoutils/utils.go [hunk 7]
```diff
@@ -66,62 +51,15 @@ func MarshalBytesEmitZeroValues(pb proto.Message) ([]byte, error) {
 	return buf.Bytes(), err
 }
 
-func StructPbToGogo(structuredData *structpb.Struct) (*types.Struct, error) {
```
[the remark this answers is not in the document]
Comment (EItanya) [reply pkg/utils/protoutils/utils.go:69]: No great reason, we need to consolidate

File: install/test/helm_test.go [hunk 8]
```diff
@@ -908,18 +909,26 @@ var _ = Describe("Helm Test", func() {
...
-							Expect(gateway1.UseProxyProto).To(Equal(&types.BoolValue{
+							Expect(gateway1.UseProxyProto).To(test_matchers.MatchProto(&wrappers.BoolValue{
 								Value: true,
 							}))
-							Expect(httpGateway.VirtualServices).To(Equal(vsList))
+							msgList = []proto.Message{}
+							for _, v := range vsList {
+								msgList = append(msgList, v)
+							}
+							Expect(httpGateway.VirtualServices).To(test_matchers.ConistOfProtos(msgList...))
```
[the remark this answers is not in the document]
Comment (EItanya) [reply install/test/helm_test.go:931]: Why, it's supposed to read like english, `Expect <values> To ConsistOfProtos<other values>`

File: install/test/fixtures/envoy_config/bootstrap_extensions.yaml [hunk 9]
```diff
@@ -135,7 +135,7 @@ bootstrap_extensions:
     '@type': type.googleapis.com/envoy.extensions.wasm.v3.WasmService
     config:
       name: my_plugin
-      vm_config:
+      inline_vm_config:
```
[the remark this answers is not in the document]
Comment (EItanya) [reply install/test/fixtures/envoy_config/bootstrap_extensions.yaml]: Good point, this was deprecated, I fixed it

File: projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto (hunk 5 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto]: Switched

File: projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto (hunk 4 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/filters/http/wasm/v3/wasm.proto]: deleted

**Review by kdorosh — APPROVED** [review]

**Pull request merged**